controllers/configuration/backend/s3.go

```diff
@@ -0,0 +1,190 @@
+/*
+Copyright 2021 The KubeVela Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package backend
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/s3"
+	"github.com/aws/aws-sdk-go/service/s3/s3iface"
+	"github.com/oam-dev/terraform-controller/api/v1beta2"
+	"github.com/pkg/errors"
+)
+
+const (
+	s3AccessKey    = "AWS_ACCESS_KEY_ID"
+	s3SecretKey    = "AWS_SECRET_ACCESS_KEY"
+	s3SessionToken = "AWS_SESSION_TOKEN"
+)
+
+// S3Backend is used to interact with the Terraform s3 backend
+type S3Backend struct {
+	client    s3iface.S3API
+	AccessKey string
+	SecretKey string
+	Token     string
+	Region    string
+	Key       string
+	Bucket    string
+}
+
+func newS3BackendFromInline(ctx k8sContext, backendConfig *ParsedBackendConfig, optionSource *OptionSource) (Backend, error) {
```

Comment:
What are the benefits compared to unmarshalling the inlined HCL? The latter one is much simple compared to the current implementation.